// @keystonejs/fields/types/Url/views/Field/dist/fields.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@emotion/core');
var fields = require('@arch-ui/fields');
var input = require('@arch-ui/input');

/** @jsx jsx */

const UrlField = ({
  onChange,
  autoFocus,
  field,
  value = '',
  errors,
  isDisabled
}) => {
  const handleChange = event => {
    onChange(event.target.value);
  };

  const htmlID = `ks-input-${field.path}`;
  const canRead = errors.every(error => !(error instanceof Error && error.name === 'AccessDeniedError'));
  const error = errors.find(error => error instanceof Error && error.name === 'AccessDeniedError');
  return core.jsx(fields.FieldContainer, null, core.jsx(fields.FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), core.jsx(fields.FieldDescription, {
    text: field.adminDoc
  }), core.jsx(fields.FieldInput, null, core.jsx(input.Input, {
    autoComplete: "off",
    autoFocus: autoFocus,
    required: field.isRequired,
    type: "url",
    value: canRead ? value : undefined,
    placeholder: canRead ? undefined : error.message,
    onChange: handleChange,
    id: htmlID,
    disabled: isDisabled
  })));
};

exports.default = UrlField;
