// @keystonejs/fields/types/Select/views/Controller/dist/fields.esm.js
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread2';
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import FieldController from '../../../../../Controller/dist/fields.esm.js';
import '@babel/runtime/helpers/classCallCheck';
import 'lodash.isequal';

class SelectController extends FieldController {
  constructor(_ref, ...args) {
    let {
      defaultValue = null
    } = _ref,
        config = _objectWithoutProperties(_ref, ["defaultValue"]);

    super(_objectSpread(_objectSpread({}, config), {}, {
      defaultValue
    }), ...args);

    _defineProperty(this, "getFilterGraphQL", ({
      value: {
        inverted,
        options
      }
    }) => {
      if (!options.length) {
        return '';
      }

      const isMulti = options.length > 1;
      let key = this.path;

      if (isMulti && inverted) {
        key = `${this.path}_not_in`;
      } else if (isMulti) {
        key = `${this.path}_in`;
      } else if (inverted) {
        key = `${this.path}_not`;
      }

      const value = isMulti ? options.map(x => x.value) : options[0].value;
      return {
        [key]: value
      };
    });

    _defineProperty(this, "getFilterLabel", () =>
    /*{ value }*/
    {
      return this.label;
    });

    _defineProperty(this, "formatFilter", ({
      value
    }) => {
      if (!value.options.length) {
        return value.inverted ? `${this.label} is set` : `${this.label} has no value`;
      }

      if (value.options.length > 1) {
        const values = value.options.map(i => i.label).join(', ');
        return value.inverted ? `${this.label} is not in [${values}]` : `${this.label} is in [${values}]`;
      }

      const optionLabel = value.options[0].label;
      return value.inverted ? `${this.label} is not ${optionLabel}` : `${this.label} is ${optionLabel}`;
    });

    _defineProperty(this, "getFilterValue", value => {
      return value && value.options && value.options.length ? value : undefined;
    });

    _defineProperty(this, "getFilterTypes", () => [{
      type: 'is',
      label: 'Matches',
      getInitialValue: () => ({
        inverted: false,
        options: []
      })
    }]);

    this.options = config.options;
    this.dataType = config.dataType;
  }

}

export default SelectController;
