// @arch-ui/common/dist/common.esm.js
import { borderRadius, gridSize } from '@arch-ui/theme';

const XL_DEVICE = 1199;
const LG_DEVICE = 991;
const MD_DEVICE = 768;
const SM_DEVICE = 575;
const smOnly = `@media (max-width: ${SM_DEVICE}px)`;
const mdOnly = `@media (min-width: ${MD_DEVICE}px) and (max-width: ${LG_DEVICE}px)`;
const lgOnly = `@media (min-width: ${LG_DEVICE}px) and (max-width: ${XL_DEVICE}px)`;
const xlOnly = `@media (min-width: ${XL_DEVICE}px)`;
const mdUp = `@media (min-width: ${MD_DEVICE}px)`;
const lgUp = `@media (min-width: ${LG_DEVICE}px)`;
const mdDown = `@media (max-width: ${MD_DEVICE}px)`;
const lgDown = `@media (max-width: ${LG_DEVICE}px)`;
const xlDown = `@media (max-width: ${XL_DEVICE}px)`;

var mediaQueries = /*#__PURE__*/Object.freeze({
  __proto__: null,
  smOnly: smOnly,
  mdOnly: mdOnly,
  lgOnly: lgOnly,
  xlOnly: xlOnly,
  mdUp: mdUp,
  lgUp: lgUp,
  mdDown: mdDown,
  lgDown: lgDown,
  xlDown: xlDown
});

const uniformHeight = {
  appearance: 'none',
  background: 'none',
  border: '1px solid transparent',
  borderRadius: borderRadius,
  boxSizing: 'border-box',
  fontFamily: 'inherit',
  fontSize: '1.0rem',
  margin: 0,
  minWidth: 1,
  padding: `${gridSize}px ${gridSize * 1.5}px`,
  transition: 'box-shadow 100ms linear',
  verticalAlign: 'middle',
  whiteSpace: 'nowrap'
};

export { mediaQueries, uniformHeight };
