// @keystonejs/fields/types/Checkbox/views/Field/dist/fields.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@emotion/core');
var fields = require('@arch-ui/fields');
var controls = require('@arch-ui/controls');

/** @jsx jsx */

const CheckboxField = ({
  onChange,
  autoFocus,
  field,
  value,
  errors,
  isDisabled
}) => {
  const handleChange = event => {
    onChange(event.target.checked);
  };

  const checked = value || false;
  const htmlID = `ks-input-${field.path}`;
  return core.jsx(fields.FieldContainer, null, core.jsx(fields.FieldDescription, {
    text: field.adminDoc
  }), core.jsx(fields.FieldInput, {
    css: {
      height: 35,
      alignItems: 'center'
    }
  }, core.jsx(controls.CheckboxPrimitive, {
    autoFocus: autoFocus,
    checked: checked,
    onChange: handleChange,
    id: htmlID,
    isDisabled: isDisabled
  }), core.jsx(fields.FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors,
    css: {
      padding: '4px',
      fontSize: '1rem',
      width: '100%'
    }
  })));
};

exports.default = CheckboxField;
