// @keystonejs/fields/types/Decimal/views/Field/dist/fields.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var fields = require('@arch-ui/fields');
var input = require('@arch-ui/input');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

const DecimalField = ({
  onChange,
  autoFocus,
  field,
  value,
  errors,
  isDisabled
}) => {
  const handleChange = event => {
    const value = event.target.value;
    onChange(value.replace(/[^0-9.,]+/g, ''));
  };

  const valueToString = value => {
    // Make the value a string to prevent loss of accuracy and precision.
    if (typeof value === 'string') {
      return value;
    } else if (typeof value === 'number') {
      return String(value);
    } else {
      // If it is neither string nor number then it must be empty.
      return '';
    }
  };

  const {
    // currency,
    // digits,
    symbol
  } = field.config;
  const htmlID = `ks-input-${field.path}`;
  return /*#__PURE__*/React__default['default'].createElement(fields.FieldContainer, null, /*#__PURE__*/React__default['default'].createElement(fields.FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), /*#__PURE__*/React__default['default'].createElement(fields.FieldDescription, {
    text: field.adminDoc
  }), /*#__PURE__*/React__default['default'].createElement(fields.FieldInput, null, symbol && /*#__PURE__*/React__default['default'].createElement(fields.Currency, null, symbol), /*#__PURE__*/React__default['default'].createElement(input.Input, {
    autoComplete: "off",
    autoFocus: autoFocus,
    required: field.isRequired,
    type: "text",
    value: valueToString(value),
    onChange: handleChange,
    id: htmlID,
    disabled: isDisabled
  })));
};

exports.default = DecimalField;
