// @arch-ui/button/dist/button.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime/helpers/extends');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var react = require('react');
var core = require('@emotion/core');
var styled = require('@emotion/styled');
var loading = require('@arch-ui/loading');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var reactRouterDom = require('react-router-dom');
var reactPseudoState = require('react-pseudo-state');
var theme = require('@arch-ui/theme');
var common = require('@arch-ui/common');
var colorUtils = require('@arch-ui/color-utils');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var styled__default = /*#__PURE__*/_interopDefault(styled);

const boldAppearance = {
  default: {
    bg: '#fff',
    border: theme.colors.N20,
    focusRing: theme.colors.primary,
    text: theme.colors.text
  },
  primary: {
    bg: theme.colors.primary,
    border: theme.colors.primary,
    focusRing: null,
    text: '#fff'
  },
  create: {
    bg: theme.colors.create,
    border: theme.colors.create,
    focusRing: null,
    text: '#fff'
  },
  danger: {
    bg: theme.colors.danger,
    border: theme.colors.danger,
    focusRing: null,
    text: '#fff'
  },
  warning: {
    bg: theme.colors.warning,
    border: theme.colors.warning,
    focusRing: null,
    text: '#fff'
  }
};
const ghostAppearance = {
  default: {
    border: theme.colors.N20,
    text: theme.colors.N60
  },
  primary: {
    border: theme.colors.B.L50,
    text: theme.colors.primary
  },
  create: {
    border: theme.colors.G.L50,
    text: theme.colors.create
  },
  danger: {
    border: theme.colors.R.L50,
    text: theme.colors.danger
  },
  warning: {
    border: theme.colors.Y.L30,
    text: theme.colors.warning
  }
};
const subtleAppearance = {
  default: {
    text: theme.colors.N40,
    textHover: theme.colors.text
  },
  primary: {
    text: theme.colors.N40,
    textHover: theme.colors.primary
  },
  create: {
    text: null,
    textHover: null
  },
  warning: {
    text: theme.colors.N40,
    textHover: theme.colors.danger
  },
  danger: {
    text: theme.colors.danger,
    textHover: theme.colors.danger
  }
};
const nuanceAppearance = {
  default: {
    text: theme.colors.text
  },
  primary: {
    text: theme.colors.primary
  },
  create: {
    text: null
  },
  warning: {
    text: theme.colors.danger
  },
  danger: {
    text: theme.colors.danger
  }
};
function makeSubtleVariant({
  appearance
}) {
  const {
    text,
    textHover
  } = subtleAppearance[appearance];
  return {
    color: text,
    fontWeight: null,
    ':hover, :focus': {
      color: textHover,
      textDecoration: 'underline'
    }
  };
}
function makeNuanceVariant({
  appearance,
  isDisabled
}) {
  const {
    text
  } = nuanceAppearance[appearance];
  return {
    color: text,
    fontWeight: 'normal',
    ':hover, :focus': makeGhostVariant({
      appearance,
      isDisabled
    })
  };
} // Ghost
// ------------------------------

function makeGhostVariant({
  appearance,
  isDisabled
}) {
  const {
    border,
    text
  } = ghostAppearance[appearance];
  return {
    border: '1px solid',
    borderColor: border,
    color: text,
    fontWeight: 'normal',
    opacity: isDisabled ? 0.5 : null,
    ':hover, :focus': {
      backgroundColor: colorUtils.alpha(border, 0.1),
      borderColor: colorUtils.darken(border, 10)
    },
    ':active': {
      color: colorUtils.darken(text, 10),
      borderColor: colorUtils.darken(border, 20),
      backgroundColor: colorUtils.alpha(border, 0.2)
    }
  };
} // Bold
// ------------------------------

function makeBoldVariant({
  appearance,
  isDisabled,
  isActive,
  isHover,
  isFocus,
  isSelected
}) {
  const {
    bg,
    border,
    focusRing,
    text
  } = boldAppearance[appearance];
  const activeBg = appearance === 'default' ? theme.colors.N10 : colorUtils.darken(bg, 12);
  const hoverAndFocus = isHover || isFocus ? {
    borderColor: border,
    background: bg
  } : null;
  const hoverStyles = isHover ? _objectSpread(_objectSpread({}, hoverAndFocus), {}, {
    boxShadow: '0 1px 2px rgba(0, 0, 0, 0.12)'
  }) : null;
  const focusStyles = isFocus && !isDisabled ? _objectSpread(_objectSpread({}, hoverAndFocus), {}, {
    borderColor: focusRing,
    boxShadow: `0 0 0 3px ${colorUtils.alpha(focusRing || bg, 0.2)}`
  }) : null;
  const activeStyles = isActive || isSelected ? {
    background: activeBg,
    borderColor: border
  } : null;
  return _objectSpread(_objectSpread(_objectSpread({
    backgroundColor: bg,
    backgroundRepeat: 'repeat-x',
    borderColor: border,
    color: text,
    fontWeight: 'bold'
  }, hoverStyles), focusStyles), activeStyles);
}

const SPACING_OPTION = {
  comfortable: `${theme.gridSize}px ${theme.gridSize * 1.5}px`,
  cozy: '2px 6px',
  cramped: '1px 2px'
};

function makeVariant({
  appearance = 'default',
  isActive,
  isBlock,
  isHover,
  isFocus,
  isDisabled,
  isSelected,
  variant = 'bold',
  spacing = 'comfortable'
}) {
  let variantStyles;
  const config = {
    appearance,
    isDisabled,
    isActive,
    isHover,
    isFocus,
    isSelected
  };

  if (variant === 'subtle') {
    variantStyles = makeSubtleVariant(config);
  } else if (variant === 'nuance') {
    variantStyles = makeNuanceVariant(config);
  } else if (variant === 'bold') {
    variantStyles = makeBoldVariant(config);
  } else if (variant === 'ghost') {
    variantStyles = makeGhostVariant(config);
  }

  return _objectSpread(_objectSpread({}, common.uniformHeight), {}, {
    cursor: isDisabled ? 'default' : 'pointer',
    display: isBlock ? 'block' : 'inline-block',
    opacity: isDisabled ? 0.66 : null,
    outline: 0,
    padding: SPACING_OPTION[spacing],
    pointerEvents: isDisabled ? 'none' : null,
    textAlign: 'center',
    touchAction: 'manipulation',
    // Disables "double-tap to zoom" for mobile; removes delay on click events
    userSelect: 'none',
    width: isBlock ? '100%' : null,
    // override possible anchor styles
    ':hover': {
      textDecoration: 'none'
    }
  }, variantStyles);
} // remove props that will create react DOM warnings


function ButtonElementComponent(props, ref) {
  const {
    isDisabled,
    isActive,
    isFocus,
    isHover,
    isSelected,
    focusOrigin
  } = props,
        rest = _objectWithoutProperties(props, ["isDisabled", "isActive", "isFocus", "isHover", "isSelected", "focusOrigin"]);

  const variant = makeVariant(props);

  if (rest.to) {
    return core.jsx(reactRouterDom.Link, _extends({
      ref: ref,
      css: variant
    }, rest));
  }

  if (rest.href) {
    return core.jsx("a", _extends({
      css: variant
    }, rest, {
      ref: ref
    }));
  }

  return core.jsx("button", _extends({
    type: "button",
    disabled: isDisabled,
    css: variant,
    ref: ref
  }, rest));
}

const ButtonElement = /*#__PURE__*/react.forwardRef(ButtonElementComponent);
const Button = reactPseudoState.withPseudoState(ButtonElement);

const LoadingButtonInner = styled__default['default'].div({
  position: 'relative'
});
const LoadingIndicatorWrapper = styled__default['default'].div({
  left: '50%',
  position: 'absolute',
  top: '50%',
  transform: 'translate(-50%, -50%)'
});

function getAppearance(appearance) {
  if (appearance === 'default') return 'dark';
  return 'inverted';
}

function LoadingButtonComponent(_ref, ref) {
  let {
    children,
    indicatorVariant,
    isLoading
  } = _ref,
      props = _objectWithoutProperties(_ref, ["children", "indicatorVariant", "isLoading"]);

  const appearance = getAppearance(props.appearance || 'default');
  const textCSS = isLoading ? {
    visibility: 'hidden'
  } : null;
  const isSpinner = indicatorVariant === 'spinner';
  return core.jsx(Button, _extends({
    ref: ref,
    variant: "bold"
  }, props), core.jsx(LoadingButtonInner, null, isLoading ? core.jsx(LoadingIndicatorWrapper, null, isSpinner ? core.jsx(loading.LoadingSpinner, {
    appearance: appearance,
    size: 16
  }) : core.jsx(loading.LoadingIndicator, {
    appearance: appearance,
    size: 4
  })) : null, core.jsx("span", {
    css: textCSS
  }, children)));
} // Export


const LoadingButton = /*#__PURE__*/react.forwardRef(LoadingButtonComponent);

const IconButton = /*#__PURE__*/react.forwardRef((_ref, ref) => {
  let {
    children,
    icon: Icon,
    iconSize = 16
  } = _ref,
      props = _objectWithoutProperties(_ref, ["children", "icon", "iconSize"]);

  return core.jsx(Button, _extends({
    ref: ref
  }, props), core.jsx("span", {
    css: {
      display: 'flex',
      alignItems: 'center'
    }
  }, core.jsx(Icon, {
    size: iconSize,
    css: children ? {
      marginRight: '0.5em'
    } : null
  }), children));
});

exports.Button = Button;
exports.IconButton = IconButton;
exports.LoadingButton = LoadingButton;
