// @keystonejs/fields/types/Select/views/Filter/dist/fields.esm.js
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread2';
import _extends from '@babel/runtime/helpers/esm/extends';
import { jsx } from '@emotion/core';
import { Fragment } from 'react';
import Select from '@arch-ui/select';
import { Options, OptionPrimitive, CheckMark } from '@arch-ui/options';
import { RadioGroup, Radio } from '@arch-ui/filters';
import { gridSize } from '@arch-ui/theme';

const EventCatcher = props => jsx("div", _extends({
  onClick: e => {
    e.preventDefault();
    e.stopPropagation();
  }
}, props));

const SelectWrapper = props => jsx("div", _extends({
  css: {
    marginTop: gridSize * 2
  }
}, props));

const SelectFilterView = ({
  innerRef,
  field,
  value,
  onChange
}) => {
  const handleRadioChange = newValue => {
    const inverted = newValue === 'does_match' ? false : true;
    onChange(_objectSpread(_objectSpread({}, value), {}, {
      inverted
    }));
  };

  const handleSelectChange = newValue => {
    const options = [...(newValue || [])]; // ensure consistent data shape

    onChange(_objectSpread(_objectSpread({}, value), {}, {
      options
    }));
  };

  const radioValue = value.inverted ? 'does_not_match' : 'does_match';
  const selectProps = {
    innerRef: innerRef,
    onChange: handleSelectChange,
    options: field.options,
    placeholder: 'Select...',
    value: value.options,
    isMulti: true
  };
  return jsx(Fragment, null, jsx(RadioGroup, {
    onChange: handleRadioChange,
    value: radioValue
  }, jsx(Radio, {
    value: "does_match"
  }, "Matches"), jsx(Radio, {
    value: "does_not_match"
  }, "Does not match")), jsx(SelectWrapper, null, field.options.length > 8 ? jsx(EventCatcher, null, jsx(Select, _extends({
    menuPortalTarget: document.body
  }, selectProps, {
    components: {
      Option: CheckMarkOptionSelect
    }
  }))) : jsx(Options, _extends({
    displaySearch: false
  }, selectProps, {
    components: {
      Option: CheckMarkOption
    }
  }))));
};

const CheckMarkOption = (_ref) => {
  let {
    children
  } = _ref,
      props = _objectWithoutProperties(_ref, ["children"]);

  return jsx(OptionPrimitive, props, jsx("span", null, children), jsx(CheckMark, {
    isFocused: props.isFocused,
    isSelected: props.isSelected
  }));
}; // TODO: figure out a better way to give options padding in the the dropdown


const CheckMarkOptionSelect = props => jsx(CheckMarkOption, _extends({}, props, {
  css: {
    paddingLeft: `${gridSize}px`,
    paddingRight: `${gridSize}px`
  }
}));

export default SelectFilterView;
