// @arch-ui/lozenge/dist/lozenge.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var styled = require('@emotion/styled');
var theme = require('@arch-ui/theme');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var styled__default = /*#__PURE__*/_interopDefault(styled);

const boldBackgroundColor = {
  default: theme.colors.N60,
  primary: theme.colors.primary,
  danger: theme.colors.danger,
  create: theme.colors.create,
  warning: theme.colors.warning
};
const boldTextColor = {
  default: 'white',
  primary: 'white',
  danger: 'white',
  create: 'white',
  warning: 'white'
};
const subtleBackgroundColor = {
  default: theme.colors.N05,
  primary: theme.colors.B.L90,
  danger: theme.colors.R.L90,
  create: theme.colors.G.L90,
  warning: theme.colors.Y.L90
};
const subtleBorderColor = {
  default: theme.colors.N10,
  primary: theme.colors.B.L80,
  danger: theme.colors.R.L80,
  create: theme.colors.G.L80,
  warning: theme.colors.Y.L80
};
const subtleTextColor = {
  default: theme.colors.N50,
  primary: theme.colors.B.L30,
  danger: theme.colors.R.L30,
  create: theme.colors.G.L30,
  warning: theme.colors.Y.L30
};
const Lozenge = styled__default['default'].div(({
  appearance = 'default',
  variant = 'subtle',
  crop
}) => {
  const isClipRight = crop === 'right';
  const cropStyles = crop ? {
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    // clip from beginning of string
    direction: isClipRight ? 'rtl' : null,
    textAlign: isClipRight ? 'left' : null
  } : null;
  return _objectSpread({
    border: '1px solid transparent',
    backgroundColor: variant === 'bold' ? boldBackgroundColor[appearance] : subtleBackgroundColor[appearance],
    borderColor: variant === 'bold' ? null : subtleBorderColor[appearance],
    boxSizing: 'border-box',
    color: variant === 'bold' ? boldTextColor[appearance] : subtleTextColor[appearance],
    borderRadius: theme.borderRadius,
    display: 'inline-block',
    fontSize: '0.85em',
    maxWidth: '100%',
    minWidth: 1,
    padding: `${theme.gridSize / 2}px ${theme.gridSize}px`
  }, cropStyles);
});

exports.Lozenge = Lozenge;
