// react-day-picker/lib/components/DayPicker/defaults/defaultProps.js
import english from 'date-fns/locale/en-US';
import { Caption } from '../../Caption/Caption';
import { Day } from '../../Day/Day';
import { Navigation } from '../../Navigation';
import { NextIcon } from '../../NextIcon';
import { PrevIcon } from '../../PrevIcon';
import { Row } from '../../Row';
import { defaultClassNames } from './defaultClassNames';
import { defaultLabels } from './defaultLabels';
import { defaultModifiersClassNames } from './defaultModifiersClassNames';
import { formatCaption } from '../utils/formatCaption';
import { formatDay } from '../utils/formatDay';
import { formatWeekdayName } from '../utils/formatWeekdayName';
import { formatWeekNumber } from '../utils/formatWeekNumber';
export var defaultProps = {
    labelsFormatters: defaultLabels,
    enableOutsideDaysClick: false,
    classNames: defaultClassNames,
    className: '',
    components: {
        Caption: Caption,
        Day: Day,
        Navigation: Navigation,
        NextIcon: NextIcon,
        PrevIcon: PrevIcon,
        Week: Row
    },
    fixedWeeks: false,
    formatCaption: formatCaption,
    formatDay: formatDay,
    formatWeekdayName: formatWeekdayName,
    formatWeekNumber: formatWeekNumber,
    initialMonth: undefined,
    locale: english,
    modifiersClassNames: defaultModifiersClassNames,
    month: new Date(),
    numberOfMonths: 1,
    pagedNavigation: false,
    reverseMonths: false,
    showCaption: true,
    showHead: true,
    showNavigation: true,
    showOutsideDays: false,
    showWeekNumber: false,
    today: new Date()
};
