// react-day-picker/lib/components/DayPicker/defaults/defaultLabels.js
import { format } from 'date-fns';
import { defaultProps } from './defaultProps';
export var defaultLabels = {
    navNextButton: function (month, props) {
        var _a;
        var formatCaption = (_a = props === null || props === void 0 ? void 0 : props.formatCaption) !== null && _a !== void 0 ? _a : defaultProps.formatCaption;
        return "Go to next month: " + formatCaption(month, props);
    },
    navPrevButton: function (month, props) {
        var _a;
        var formatCaption = (_a = props === null || props === void 0 ? void 0 : props.formatCaption) !== null && _a !== void 0 ? _a : defaultProps.formatCaption;
        return "Go to previous month: " + formatCaption(month, props);
    },
    day: function (day, _, props) {
        return format(day, 'PPPP', { locale: props.locale });
    },
    headCell: function (day, props) {
        return props.formatWeekdayName(day, { locale: props.locale }, 'cccc');
    },
    rowHead: function (n, props) {
        return "Week n. " + n;
    }
};
