// @keystonejs/app-version/dist/app-version.esm.js
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import { parseCustomAccess } from '@keystonejs/access-control';

class AppVersionProvider {
  constructor({
    version,
    access,
    schemaNames
  }) {
    _defineProperty(this, "_access", void 0);

    _defineProperty(this, "_version", void 0);

    this._access = parseCustomAccess({
      access: access,
      schemaNames,
      defaultAccess: true
    });
    this._version = version;
  }

  getTypes({}) {
    return [];
  }

  getQueries({
    schemaName
  }) {
    return this._access[schemaName] ? [`"""The version of the Keystone application serving this API."""
          appVersion: String`] : [];
  }

  getMutations({}) {
    return [];
  }

  getSubscriptions({}) {
    return [];
  }

  getTypeResolvers({}) {
    return {};
  }

  getQueryResolvers({
    schemaName
  }) {
    return this._access[schemaName] ? {
      appVersion: () => this._version
    } : {};
  }

  getMutationResolvers({}) {
    return {};
  }

  getSubscriptionResolvers({}) {
    return {};
  }

}
function appVersionMiddleware(version) {
  return (req, res, next) => {
    res.set('X-Keystone-App-Version', version);
    next();
  };
}

export { AppVersionProvider, appVersionMiddleware };
