// @arch-ui/drawer/dist/drawer.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime/helpers/extends');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var core = require('@emotion/core');
var react = require('react');
var reactDom = require('react-dom');
var styled = require('@emotion/styled');
var ScrollLock = require('react-scrolllock');
var FocusTrap = require('focus-trap-react');
var modalUtils = require('@arch-ui/modal-utils');
var theme = require('@arch-ui/theme');
var colorUtils = require('@arch-ui/color-utils');
var typography = require('@arch-ui/typography');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var styled__default = /*#__PURE__*/_interopDefault(styled);
var ScrollLock__default = /*#__PURE__*/_interopDefault(ScrollLock);
var FocusTrap__default = /*#__PURE__*/_interopDefault(FocusTrap);

// inspired by https://bitbucket.org/atlassian/atlaskit-mk-2/src/75c59d2c870d7e2489d6baea9a6837f9bd4082c4/packages/core/modal-dialog/src/components/StackConsumer.js
// TODO: investigate if this will cause problems with concurrent mode

let allStackConsumers = {
  left: [],
  right: []
};

let updateStackConsumers = slideInFrom => {
  allStackConsumers[slideInFrom].forEach(update => {
    update();
  });
}; // This hook provides the position of a modal dialog in the list of all open dialogs.
// The key behaviours are:
// - When a modal renders for the first time it takes the first stack position
// - When a modal mounts, all other modals have to adjust their position
// - When a modal unmounts, all other modals have to adjust their position


function useStackIndex(isOpen, slideInFrom) {
  let [stackIndex, setStackIndex] = react.useState(isOpen ? 0 : -1);
  react.useLayoutEffect(() => {
    if (isOpen) {
      let update = () => {
        setStackIndex(allStackConsumers[slideInFrom].indexOf(update));
      };

      allStackConsumers[slideInFrom].unshift(update);
      updateStackConsumers(slideInFrom);
      return () => {
        allStackConsumers[slideInFrom] = allStackConsumers[slideInFrom].filter(x => x !== update);
        updateStackConsumers(slideInFrom);
      };
    } else {
      setStackIndex(-1);
    }
  }, [isOpen, slideInFrom]);
  return stackIndex;
}

const innerGutter = theme.gridSize * 2; // Styled Components
// ------------------------------

const Positioner = (_ref) => {
  let {
    slideInFrom,
    width,
    stackIndex,
    style: {
      transform
    }
  } = _ref,
      style = _objectWithoutProperties(_ref.style, ["transform"]),
      props = _objectWithoutProperties(_ref, ["slideInFrom", "width", "stackIndex", "style"]);

  const stackTransforms = stackIndex <= 0 ? [] : [`translate(calc(${stackIndex * 0.3} * -9vw))`, `scale(${1 - stackIndex / 50})`];
  return core.jsx("div", _extends({
    css: _objectSpread({
      boxSizing: 'border-box',
      padding: theme.gridSize,
      display: 'flex',
      flexDirection: 'column',
      height: '100%',
      maxWidth: '90%',
      position: 'fixed',
      [slideInFrom]: 0,
      top: 0,
      width,
      zIndex: 2,
      transform: `${transform}${stackTransforms.join(' ')}`
    }, style)
  }, props));
};

const Dialog = /*#__PURE__*/react.forwardRef((_ref2, ref) => {
  let {
    component: Tag
  } = _ref2,
      props = _objectWithoutProperties(_ref2, ["component"]);

  return core.jsx(Tag, _extends({
    ref: ref,
    role: "dialog",
    css: {
      backgroundColor: 'white',
      boxShadow: theme.shadows[3],
      borderRadius: theme.borderRadius * 2,
      display: 'flex',
      flex: 1,
      flexDirection: 'column',
      // margin: gridSize,
      maxHeight: '100%'
    }
  }, props));
}); // Content

const HeadFoot = styled__default['default'].div({
  lineHeight: 1,
  margin: `0 ${innerGutter}px`,
  paddingBottom: innerGutter,
  paddingTop: innerGutter,
  // ensure that box-shadow covers body content
  position: 'relative',
  zIndex: 1
});
const Header = styled__default['default'](HeadFoot)({
  boxShadow: `0 2px 0 ${colorUtils.alpha(theme.colors.text, 0.12)}`
});
const Footer = styled__default['default'](HeadFoot)({
  boxShadow: `0 -2px 0 ${colorUtils.alpha(theme.colors.text, 0.12)}`
});
const Body = styled__default['default'].div({
  lineHeight: 1.4,
  overflowY: 'auto',
  padding: innerGutter
});

function useKeydownHandler(handler) {
  let handlerRef = react.useRef(handler);
  react.useEffect(() => {
    handlerRef.current = handler;
  });
  react.useEffect(() => {
    function handle(event) {
      handlerRef.current(event);
    }

    document.addEventListener('keydown', handle, false);
    return () => {
      document.removeEventListener('keydown', handle, false);
    };
  }, []);
}

function ModalDialogComponent({
  attachTo = typeof document !== 'undefined' ? document.body : null,
  children,
  closeOnBlanketClick = false,
  component = 'div',
  footer,
  heading,
  initialFocus,
  onClose,
  slideInFrom,
  width = 640,
  onKeyDown,
  transitionState,
  isOpen
}) {
  let stackIndex = useStackIndex(transitionState === 'entered' || transitionState === 'entering', slideInFrom);
  useKeydownHandler(event => {
    if (onKeyDown && stackIndex === 0) {
      onKeyDown(event);
    }
  });
  const dialogTitleId = react.useMemo(modalUtils.generateUEID, []);

  if (!attachTo) {
    return null;
  }

  return /*#__PURE__*/reactDom.createPortal(core.jsx(react.Fragment, null, isOpen ? core.jsx(modalUtils.Blanket, {
    style: modalUtils.fade(transitionState),
    onClick: closeOnBlanketClick ? onClose : undefined,
    isTinted: true
  }) : null, core.jsx(Positioner, {
    style: modalUtils.slideInHorizontal(transitionState, {
      slideInFrom
    }),
    slideInFrom: slideInFrom,
    width: width,
    stackIndex: stackIndex
  }, core.jsx(FocusTrap__default['default'], {
    focusTrapOptions: {
      initialFocus,
      clickOutsideDeactivates: closeOnBlanketClick
    }
  }, core.jsx(Dialog, {
    component: component,
    "aria-labelledby": dialogTitleId
  }, core.jsx(typography.A11yText, {
    id: dialogTitleId
  }, heading, " Dialog"), heading ? core.jsx(Header, null, core.jsx(typography.Title, null, heading)) : null, core.jsx(ScrollLock__default['default'], null, core.jsx(Body, null, children)), footer ? core.jsx(Footer, null, footer) : null)))), attachTo);
}

const ModalDialog = /*#__PURE__*/react.memo(ModalDialogComponent);
var index = modalUtils.withTransitionState(ModalDialog);

exports.default = index;
