// @arch-ui/typography/dist/typography.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var _extends = require('@babel/runtime/helpers/extends');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var core = require('@emotion/core');
var styled = require('@emotion/styled');
var theme = require('@arch-ui/theme');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var styled__default = /*#__PURE__*/_interopDefault(styled);

const truncate = {
  minWidth: 0,
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap'
};
const SubtleText = styled__default['default'].span({
  color: theme.colors.N60
});
const H1 = styled__default['default'].h1({
  fontSize: 28,
  fontWeight: 300,
  margin: '24px 0'
});
const PageTitle = styled__default['default'].h1({
  fontSize: '2.4rem',
  fontWeight: 900,
  marginBottom: '0.5em',
  marginTop: '1.0rem'
});
const Truncate = (_ref) => {
  let {
    as: Tag = 'div'
  } = _ref,
      props = _objectWithoutProperties(_ref, ["as"]);

  return core.jsx(Tag, _extends({
    css: truncate
  }, props));
};
const Title = (_ref2) => {
  let {
    as: Tag = 'h3',
    crop = false,
    margin = 'none'
  } = _ref2,
      props = _objectWithoutProperties(_ref2, ["as", "crop", "margin"]);

  const gutter = theme.gridSize * 3;
  const margins = {
    none: {
      margin: 0
    },
    both: {
      marginBottom: gutter,
      marginTop: gutter
    },
    bottom: {
      marginBottom: gutter,
      marginTop: 0
    },
    top: {
      marginBottom: 0,
      marginTop: gutter
    }
  };
  const offset = margins[margin];
  const cropStyles = crop ? truncate : null;
  return core.jsx(Tag, _extends({
    css: _objectSpread(_objectSpread({
      fontSize: 18,
      fontWeight: 600,
      whiteSpace: 'nowrap'
    }, cropStyles), offset)
  }, props));
};
const Kbd = styled__default['default'].kbd({
  backgroundColor: theme.colors.N05,
  border: `1px solid ${theme.colors.N20}`,
  borderRadius: 3,
  boxShadow: '0 1px 1px rgba(0, 0, 0, 0.12), 0 2px 0 0 rgba(255, 255, 255, 0.7) inset',
  display: 'inline-block',
  fontFamily: 'Monaco, monospace',
  fontSize: '0.85em',
  fontWeight: 'bold',
  lineHeight: 'inherit',
  padding: '1px 5px',
  position: 'relative',
  top: '-1px',
  verticalAlign: 'middle',
  whiteSpace: 'nowrap'
});
const A11yText = (_ref3) => {
  let {
    tag: Tag = 'span'
  } = _ref3,
      props = _objectWithoutProperties(_ref3, ["tag"]);

  return core.jsx(Tag, _extends({
    css: {
      border: 0,
      clip: 'rect(1px, 1px, 1px, 1px)',
      height: 1,
      overflow: 'hidden',
      padding: 0,
      position: 'absolute',
      whiteSpace: 'nowrap',
      width: 1
    }
  }, props));
};

exports.A11yText = A11yText;
exports.H1 = H1;
exports.Kbd = Kbd;
exports.PageTitle = PageTitle;
exports.SubtleText = SubtleText;
exports.Title = Title;
exports.Truncate = Truncate;
