// @arch-ui/badge/dist/badge.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime/helpers/extends');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var React = require('react');
var styled = require('@emotion/styled');
var theme = require('@arch-ui/theme');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefault(React);
var styled__default = /*#__PURE__*/_interopDefault(styled);

const boldBackgroundColor = {
  default: theme.colors.N60,
  inverted: theme.colors.text,
  primary: theme.colors.primary,
  created: theme.colors.create,
  warning: theme.colors.warning,
  danger: theme.colors.danger
};
const boldTextColor = {
  default: 'white',
  inverted: 'white',
  primary: 'white',
  created: 'white',
  warning: 'white',
  danger: 'white'
};
const subtleBackgroundColor = {
  default: theme.colors.N10,
  inverted: 'white',
  primary: theme.colors.B.L85,
  created: theme.colors.G.L85,
  warning: theme.colors.Y.L85,
  danger: theme.colors.R.L85
};
const subtleTextColor = {
  default: theme.colors.N70,
  inverted: theme.colors.text,
  primary: theme.colors.B.D20,
  created: theme.colors.G.D20,
  warning: theme.colors.Y.D20,
  danger: theme.colors.R.D20
};
const BadgeElement = styled__default['default'].div(({
  appearance,
  variant
}) => ({
  backgroundColor: variant === 'bold' ? boldBackgroundColor[appearance] : subtleBackgroundColor[appearance],
  borderRadius: '2em',
  boxSizing: 'border-box',
  color: variant === 'bold' ? boldTextColor[appearance] : subtleTextColor[appearance],
  display: 'inline-block',
  fontSize: 12,
  fontWeight: 500,
  lineHeight: 1,
  minWidth: '2em',
  padding: '0.25em 0.5em',
  textAlign: 'center'
}));
const Badge = /*#__PURE__*/React.memo((_ref) => {
  let {
    appearance = 'default',
    max = 99,
    value = 0,
    variant = 'subtle'
  } = _ref,
      props = _objectWithoutProperties(_ref, ["appearance", "max", "value", "variant"]);

  const getValue = ({
    value,
    max
  }) => {
    if (value < 0) return '0';
    if (max > 0 && value > max) return `${max}+`;
    return value;
  };

  return /*#__PURE__*/React__default['default'].createElement(BadgeElement, _extends({
    appearance: appearance,
    variant: variant
  }, props), getValue({
    value,
    max
  }));
});

exports.Badge = Badge;
