// @arch-ui/popout/dist/popout.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var _extends = require('@babel/runtime/helpers/extends');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var core = require('@emotion/core');
var react = require('react');
var reactDom = require('react-dom');
var theme = require('@arch-ui/theme');
var FocusTrap = require('focus-trap-react');
var modalUtils = require('@arch-ui/modal-utils');
var reactPopper = require('react-popper');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var FocusTrap__default = /*#__PURE__*/_interopDefault(FocusTrap);

const ARROW_WIDTH = 30;
const Wrapper = /*#__PURE__*/react.forwardRef((_ref, ref) => {
  let {
    left,
    top,
    width
  } = _ref,
      props = _objectWithoutProperties(_ref, ["left", "top", "width"]);

  return core.jsx("div", _extends({
    ref: ref,
    css: {
      backgroundColor: 'white',
      borderRadius: theme.borderRadius * 2,
      boxShadow: theme.shadows[2],
      marginTop: theme.gridSize * 2,
      maxHeight: '100%',
      zIndex: 200,
      width
    }
  }, props));
});
const WrapperInner = /*#__PURE__*/react.forwardRef((props, ref) => core.jsx("div", _extends({
  ref: ref,
  css: {
    position: 'relative'
  }
}, props)));
const Arrow = /*#__PURE__*/react.forwardRef(({
  style
}, ref) => core.jsx("div", {
  ref: ref,
  css: {
    height: ARROW_WIDTH,
    marginTop: '-11px',
    position: 'absolute',
    width: ARROW_WIDTH
  },
  style: style
}, core.jsx("svg", {
  viewBox: "0 0 30 30",
  style: {
    transform: 'rotate(90deg)'
  },
  focusable: "false",
  role: "presentation"
}, core.jsx("path", {
  css: {
    fill: theme.colors.text,
    fillOpacity: 0.1
  },
  d: "M8.11 6.302c1.015-.936 1.887-2.922 1.887-4.297v26c0-1.378-.868-3.357-1.888-4.297L.925 17.09c-1.237-1.14-1.233-3.034 0-4.17L8.11 6.302z"
}), core.jsx("path", {
  css: {
    fill: 'white'
  },
  d: "M8.787 7.036c1.22-1.125 2.21-3.376 2.21-5.03V0v30-2.005c0-1.654-.983-3.9-2.21-5.03l-7.183-6.616c-.81-.746-.802-1.96 0-2.7l7.183-6.614z"
}))));

const Popout = ({
  children,
  targetNode,
  contentNode,
  getModalRef,
  style,
  width = 320
}) => {
  const [arrowElement, setArrowElement] = react.useState(null);
  const {
    styles
  } = reactPopper.usePopper(targetNode, contentNode, {
    placement: 'bottom',
    modifiers: [{
      name: 'arrow',
      options: {
        element: arrowElement
      }
    }]
  });
  const attachTo = typeof document !== 'undefined' ? document.body : null;
  return attachTo ? /*#__PURE__*/reactDom.createPortal(core.jsx("div", {
    ref: getModalRef,
    style: _objectSpread(_objectSpread({}, styles.popper), {}, {
      zIndex: 2
    })
  }, core.jsx(Wrapper, {
    width: width,
    style: style // style comes from Transition

  }, core.jsx(FocusTrap__default['default'], {
    focusTrapOptions: {
      clickOutsideDeactivates: true
    }
  }, core.jsx(WrapperInner, null, core.jsx(Arrow, {
    ref: setArrowElement,
    style: styles.arrow
  }), children)))), attachTo) : null;
};

var index = modalUtils.withModalHandlers(Popout, {
  transition: modalUtils.springDown
});

exports.default = index;
