// @arch-ui/popout/dist/popout.esm.js
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread2';
import _extends from '@babel/runtime/helpers/esm/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import { jsx } from '@emotion/core';
import { forwardRef, useState } from 'react';
import { createPortal } from 'react-dom';
import { borderRadius, shadows, gridSize, colors } from '@arch-ui/theme';
import FocusTrap from 'focus-trap-react';
import { withModalHandlers, springDown } from '@arch-ui/modal-utils';
import { usePopper } from 'react-popper';

const ARROW_WIDTH = 30;
const Wrapper = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    left,
    top,
    width
  } = _ref,
      props = _objectWithoutProperties(_ref, ["left", "top", "width"]);

  return jsx("div", _extends({
    ref: ref,
    css: {
      backgroundColor: 'white',
      borderRadius: borderRadius * 2,
      boxShadow: shadows[2],
      marginTop: gridSize * 2,
      maxHeight: '100%',
      zIndex: 200,
      width
    }
  }, props));
});
const WrapperInner = /*#__PURE__*/forwardRef((props, ref) => jsx("div", _extends({
  ref: ref,
  css: {
    position: 'relative'
  }
}, props)));
const Arrow = /*#__PURE__*/forwardRef(({
  style
}, ref) => jsx("div", {
  ref: ref,
  css: {
    height: ARROW_WIDTH,
    marginTop: '-11px',
    position: 'absolute',
    width: ARROW_WIDTH
  },
  style: style
}, jsx("svg", {
  viewBox: "0 0 30 30",
  style: {
    transform: 'rotate(90deg)'
  },
  focusable: "false",
  role: "presentation"
}, jsx("path", {
  css: {
    fill: colors.text,
    fillOpacity: 0.1
  },
  d: "M8.11 6.302c1.015-.936 1.887-2.922 1.887-4.297v26c0-1.378-.868-3.357-1.888-4.297L.925 17.09c-1.237-1.14-1.233-3.034 0-4.17L8.11 6.302z"
}), jsx("path", {
  css: {
    fill: 'white'
  },
  d: "M8.787 7.036c1.22-1.125 2.21-3.376 2.21-5.03V0v30-2.005c0-1.654-.983-3.9-2.21-5.03l-7.183-6.616c-.81-.746-.802-1.96 0-2.7l7.183-6.614z"
}))));

const Popout = ({
  children,
  targetNode,
  contentNode,
  getModalRef,
  style,
  width = 320
}) => {
  const [arrowElement, setArrowElement] = useState(null);
  const {
    styles
  } = usePopper(targetNode, contentNode, {
    placement: 'bottom',
    modifiers: [{
      name: 'arrow',
      options: {
        element: arrowElement
      }
    }]
  });
  const attachTo = typeof document !== 'undefined' ? document.body : null;
  return attachTo ? /*#__PURE__*/createPortal(jsx("div", {
    ref: getModalRef,
    style: _objectSpread(_objectSpread({}, styles.popper), {}, {
      zIndex: 2
    })
  }, jsx(Wrapper, {
    width: width,
    style: style // style comes from Transition

  }, jsx(FocusTrap, {
    focusTrapOptions: {
      clickOutsideDeactivates: true
    }
  }, jsx(WrapperInner, null, jsx(Arrow, {
    ref: setArrowElement,
    style: styles.arrow
  }), children)))), attachTo) : null;
};

var index = withModalHandlers(Popout, {
  transition: springDown
});

export default index;
