// @arch-ui/alert/dist/alert.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var styled = require('@emotion/styled');
var theme = require('@arch-ui/theme');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var styled__default = /*#__PURE__*/_interopDefault(styled);

const boldBackgroundColor = {
  info: theme.colors.primary,
  success: theme.colors.create,
  warning: theme.colors.warning,
  danger: theme.colors.danger
};
const boldTextColor = {
  info: 'white',
  success: 'white',
  warning: 'white',
  danger: 'white'
};
const subtleBackgroundColor = {
  info: theme.colors.B.L80,
  success: theme.colors.G.L80,
  warning: theme.colors.Y.L80,
  danger: theme.colors.R.L80
};
const subtleTextColor = {
  info: theme.colors.B.D20,
  success: theme.colors.G.D20,
  warning: theme.colors.Y.D20,
  danger: theme.colors.R.D20
};
const Alert = styled__default['default'].div(({
  appearance = 'info',
  variant = 'subtle'
}) => ({
  backgroundColor: variant === 'bold' ? boldBackgroundColor[appearance] : subtleBackgroundColor[appearance],
  color: variant === 'bold' ? boldTextColor[appearance] : subtleTextColor[appearance],
  borderRadius: theme.borderRadius,
  display: 'flex',
  fontWeight: variant === 'bold' ? 500 : null,
  minWidth: 1,
  padding: '0.9em 1.2em',
  margin: '0.9em 0',
  '& a': {
    color: variant === 'bold' ? boldTextColor[appearance] : subtleTextColor[appearance],
    textDecoration: 'underline'
  }
}));

exports.Alert = Alert;
