// @keystonejs/fields-auto-increment/dist/fields-auto-increment.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var fields = require('@keystonejs/fields');
var adapterKnex = require('@keystonejs/adapter-knex');
var adapterPrisma = require('@keystonejs/adapter-prisma');

class AutoIncrementImplementation extends fields.Implementation {
  constructor(path, _ref = {}, context = {}) {
    let {
      gqlType,
      isUnique = true,
      access = {}
    } = _ref,
        config = _objectWithoutProperties(_ref, ["gqlType", "isUnique", "access"]);

    // Apply some field type defaults before we hand off to super; see README.md
    if (typeof access === 'object') {
      access = _objectSpread({
        create: false,
        update: false,
        delete: false
      }, access);
    } // The base implementation takes care of everything else


    super(path, _objectSpread(_objectSpread({}, config), {}, {
      isUnique,
      access
    }), context); // If no valid gqlType is supplied, default based on whether or not we're the primary key

    this.gqlType = ['ID', 'Int'].includes(gqlType) ? gqlType : this.isPrimaryKey ? 'ID' : 'Int';
  }

  get _supportsUnique() {
    return true;
  }

  gqlOutputFields() {
    return [`${this.path}: ${this.gqlType}${this.isPrimaryKey ? '!' : ''}`];
  }

  gqlOutputFieldResolvers() {
    return {
      [`${this.path}`]: item => item[this.path]
    };
  }

  gqlQueryInputFields() {
    return [...this.equalityInputFields(this.gqlType), ...this.orderingInputFields(this.gqlType), ...this.inInputFields(this.gqlType)];
  }

  gqlUpdateInputFields() {
    return [`${this.path}: ${this.gqlType}`];
  }

  gqlCreateInputFields() {
    return [`${this.path}: ${this.gqlType}`];
  }

  getBackingTypes() {
    if (this.path === 'id') {
      return {
        [this.path]: {
          optional: false,
          type: 'string'
        }
      };
    }

    return {
      [this.path]: {
        optional: true,
        type: 'number | null'
      }
    };
  }

}
class KnexAutoIncrementInterface extends adapterKnex.KnexFieldAdapter {
  constructor() {
    super(...arguments); // Default isUnique to true if not specified

    this.isUnique = typeof this.config.isUnique === 'undefined' ? true : !!this.config.isUnique;
    this.isIndexed = !!this.config.isIndexed && !this.config.isUnique;
  } // Override isNotNullable defaulting logic; default to true if not specified


  get isNotNullable() {
    if (this._isNotNullable) return this._isNotNullable;
    return this._isNotNullable = !!(typeof this.knexOptions.isNotNullable === 'undefined' ? true : this.knexOptions.isNotNullable);
  }

  addToTableSchema(table) {
    // The knex `increments()` schema building function always uses the column as the primary key
    // If not isPrimaryKey use a raw `serial` instead
    // This will only work on PostgreSQL; see README.md
    if (this.field.isPrimaryKey) {
      // Fair to say primary keys are always non-nullable and uniqueness is implied by primary()
      table.increments(this.path).notNullable(); // TODO: Warning on invalid primary key config options?
    } else {
      const column = table.specificType(this.path, 'serial');
      if (this.isUnique) column.unique();else if (this.isIndexed) column.index();
      if (this.isNotNullable) column.notNullable();
    }
  }

  addToForeignTableSchema(table, {
    path,
    isUnique,
    isIndexed,
    isNotNullable
  }) {
    if (!this.field.isPrimaryKey) {
      throw `Can't create foreign key '${path}' on table "${table._tableName}"; ` + `'${this.path}' on list '${this.field.listKey}' as is not the primary key.`;
    }

    const column = table.integer(path).unsigned();
    if (isUnique) column.unique();else if (isIndexed) column.index();
    if (isNotNullable) column.notNullable();
  }

  getQueryConditions(dbPath) {
    return _objectSpread(_objectSpread(_objectSpread({}, this.equalityConditions(dbPath)), this.orderingConditions(dbPath)), this.inConditions(dbPath));
  }

}
class PrismaAutoIncrementInterface extends adapterPrisma.PrismaFieldAdapter {
  constructor() {
    super(...arguments); // Default isUnique to true if not specified

    this.isUnique = typeof this.config.isUnique === 'undefined' ? true : !!this.config.isUnique;
    this.isIndexed = !!this.config.isIndexed && !this.config.isUnique;
  }

  getPrismaSchema() {
    return [this._schemaField({
      type: 'Int',
      extra: '@default(autoincrement())'
    })];
  }

  gqlToPrisma(value) {
    // If we're an ID type then we'll be getting strings from GQL
    return Number(value); // console.log(this.field.gqlType);
    // return this.field.gqlType === 'ID' ? Number(value) : value;
  }

  equalityConditions(dbPath, f) {
    return {
      [this.path]: value => ({
        [dbPath]: f(value)
      }),
      [`${this.path}_not`]: value => ({
        NOT: {
          [this.path]: f(value)
        }
      })
    };
  }

  inConditions(dbPath, f) {
    return {
      [`${this.path}_in`]: value => value.includes(null) ? {
        [dbPath]: {
          in: f(value.filter(x => x !== null))
        }
      } : {
        [dbPath]: {
          in: f(value)
        }
      },
      [`${this.path}_not_in`]: value => value.includes(null) ? {
        AND: [{
          NOT: {
            [dbPath]: {
              in: f(value.filter(x => x !== null))
            }
          }
        }]
      } : {
        NOT: {
          [dbPath]: {
            in: f(value)
          }
        }
      }
    };
  }

  getQueryConditions(dbPath) {
    return _objectSpread(_objectSpread(_objectSpread({}, this.equalityConditions(dbPath, x => Number(x) || -1)), this.orderingConditions(dbPath, x => Number(x) || -1)), this.inConditions(dbPath, x => x.map(xx => Number(xx) || -1)));
  }

}

const AutoIncrement = {
  type: 'AutoIncrement',
  implementation: AutoIncrementImplementation,
  views: {
    Controller: fields.Integer.views.Controller,
    Field: fields.Integer.views.Field,
    Filter: fields.Integer.views.Filter
  },
  adapters: {
    knex: KnexAutoIncrementInterface,
    prisma: PrismaAutoIncrementInterface
  },
  primaryKeyDefaults: {
    knex: {
      // Uniqueness, non-nullability and GraphQL type are implied
      getConfig: () => ({
        type: AutoIncrement
      })
    },
    prisma: {
      // Uniqueness, non-nullability and GraphQL type are implied
      getConfig: () => ({
        type: AutoIncrement
      })
    }
  }
};

exports.AutoIncrement = AutoIncrement;
