// @keystonejs/fields/types/Uuid/views/Field/dist/fields.esm.js
import { jsx } from '@emotion/core';
import { FieldContainer, FieldLabel, FieldDescription, FieldInput } from '@arch-ui/fields';
import { Input } from '@arch-ui/input';

/** @jsx jsx */

const UuidField = ({
  onChange,
  autoFocus,
  field,
  errors,
  value = '',
  isDisabled
}) => {
  const handleChange = event => {
    onChange(event.target.value);
  };

  const htmlID = `ks-input-${field.path}`;
  const canRead = errors.every(error => !(error instanceof Error && error.name === 'AccessDeniedError'));
  const error = errors.find(error => error instanceof Error && error.name === 'AccessDeniedError');
  return jsx(FieldContainer, null, jsx(FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), jsx(FieldDescription, {
    text: field.adminDoc
  }), jsx(FieldInput, null, jsx(Input, {
    autoComplete: "off",
    autoFocus: autoFocus,
    required: field.isRequired,
    type: "text",
    value: canRead ? value : undefined,
    placeholder: canRead ? undefined : error.message,
    onChange: handleChange,
    id: htmlID,
    disabled: isDisabled
  })));
};

export default UuidField;
