// @arch-ui/filters/dist/filters.esm.js
import _extends from '@babel/runtime/helpers/esm/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import { jsx } from '@emotion/core';
import { Children } from 'react';
import { CheckboxGroup as CheckboxGroup$1, Checkbox as Checkbox$1, RadioGroup as RadioGroup$1, Radio as Radio$1 } from 'react-radios';
import { colors } from '@arch-ui/theme';
import { CheckboxPrimitive, RadioPrimitive } from '@arch-ui/controls';
import { FlexGroup } from '@arch-ui/layout';

const ControlLabel = (_ref) => {
  let props = _objectWithoutProperties(_ref, ["isChecked", "isDisabled"]);

  const type = Children.toArray(props.children)[0].props.type;
  const borderRadius = type === 'checkbox' ? 3 : '2em';
  return jsx("label", _extends({
    css: {
      alignItems: 'center',
      border: `1px solid ${colors.N10}`,
      borderRadius,
      display: 'flex',
      fontSize: '0.75em',
      fontWeight: 500,
      lineHeight: 1,
      transition: 'border-color 150ms linear',
      width: '100%',
      userSelect: 'none',
      ':hover, :focus': {
        borderColor: colors.N20
      },
      ':active': {
        backgroundColor: colors.N05
      }
    }
  }, props));
};

const StretchGroup = props => jsx(FlexGroup, _extends({
  stretch: true
}, props)); // checkbox


const CheckboxGroup = props => jsx(CheckboxGroup$1, _extends({
  component: StretchGroup
}, props));

const ButtonCheckbox = props => jsx(CheckboxPrimitive, _extends({
  components: {
    Label: ControlLabel
  }
}, props));

const Checkbox = props => jsx(Checkbox$1, _extends({
  component: ButtonCheckbox
}, props)); // radio

const RadioGroup = props => jsx(RadioGroup$1, _extends({
  component: StretchGroup
}, props));

const ButtonRadio = props => jsx(RadioPrimitive, _extends({
  components: {
    Label: ControlLabel
  }
}, props));

const Radio = props => jsx(Radio$1, _extends({
  component: ButtonRadio
}, props));

export { Checkbox, CheckboxGroup, Radio, RadioGroup };
