// @arch-ui/navbar/dist/navbar.esm.js
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread2';
import styled from '@emotion/styled';
import { gridSize, colors } from '@arch-ui/theme';
import _extends from '@babel/runtime/helpers/esm/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import React, { createElement } from 'react';
import { Link } from 'react-router-dom';

const ItemElement = (_ref) => {
  let {
    as: Tag,
    isSelected,
    mouseIsOverNav
  } = _ref,
      props = _objectWithoutProperties(_ref, ["as", "isSelected", "mouseIsOverNav"]);

  if (Tag) return /*#__PURE__*/createElement(Tag, props);
  if (props.to) return /*#__PURE__*/createElement(Link, props);
  if (props.href) return /*#__PURE__*/createElement("a", props);
  return /*#__PURE__*/createElement("button", _extends({
    type: "button"
  }, props));
};

const PRIMARY_NAV_GUTTER = gridSize * 2;
const NavGroupIcons = styled.div({
  alignItems: 'center',
  alignSelf: 'stretch',
  display: 'flex',
  justifyContent: 'space-between'
});
const PrimaryNav = styled.nav({
  backgroundColor: colors.page,
  boxSizing: 'border-box',
  display: 'flex',
  flexDirection: 'column',
  height: '100vh',
  position: 'fixed',
  zIndex: 2
});
const PrimaryNavScrollArea = styled.div(({
  hasScroll,
  isBottom,
  isScrollable
}) => {
  const divider = {
    backgroundColor: 'rgba(9, 30, 66, 0.1)',
    content: '" "',
    height: 2,
    left: PRIMARY_NAV_GUTTER,
    right: PRIMARY_NAV_GUTTER,
    position: 'absolute'
  };
  const before = hasScroll ? _objectSpread(_objectSpread({}, divider), {}, {
    top: 0
  }) : null;
  const after = isScrollable && !isBottom ? _objectSpread(_objectSpread({}, divider), {}, {
    bottom: 0
  }) : null;
  return {
    boxSizing: 'border-box',
    flex: 1,
    overflowY: 'auto',
    overflowX: 'hidden',
    minWidth: 140,
    paddingBottom: PRIMARY_NAV_GUTTER,
    width: '100%',
    ':before': before,
    ':after': after
  };
});
const BrandItem = styled.h2({
  fontSize: 18,
  fontWeight: 500,
  margin: 0,
  paddingBottom: PRIMARY_NAV_GUTTER
});
const PrimaryNavItem = styled(ItemElement)(({
  depth,
  isSelected,
  mouseIsOverNav
}) => {
  const selectedStyles = isSelected ? {
    '&, :hover, :active, :focus': {
      ':after': {
        backgroundColor: colors.primary
      }
    }
  } : {};
  return _objectSpread({
    border: 0,
    borderRight: '1px solid transparent',
    color: isSelected ? colors.N90 : mouseIsOverNav ? colors.N70 : colors.N40,
    display: 'block',
    marginBottom: 2,
    overflow: 'hidden',
    padding: `${gridSize * 1.5}px`,
    paddingLeft: depth ? PRIMARY_NAV_GUTTER * depth : PRIMARY_NAV_GUTTER,
    position: 'relative',
    textDecoration: 'none',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    transition: 'color 110ms',
    fontWeight: isSelected ? 'bold' : 'normal',
    flexGrow: 1,
    flexBasis: '100%',
    ':hover, :focus': {
      backgroundColor: colors.N10,
      textDecoration: 'none'
    },
    ':active': {
      backgroundColor: colors.N10
    },
    ':after': {
      borderRadius: 2,
      bottom: 2,
      content: '" "',
      pointerEvents: 'none',
      position: 'absolute',
      right: 6,
      top: 2,
      transition: 'background-color 110ms',
      width: 4
    }
  }, selectedStyles);
});
const NavIcon = styled(ItemElement)(({
  mouseIsOverNav
}) => {
  return {
    color: mouseIsOverNav ? colors.N70 : colors.N40,
    padding: PRIMARY_NAV_GUTTER,
    textDecoration: 'none',
    flexGrow: 1,
    flexBasis: '100%',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    ':hover, :focus': {
      backgroundColor: colors.N10,
      textDecoration: 'none'
    },
    ':active': {
      backgroundColor: colors.N10
    }
  };
});
const PrimaryNavHeading = styled.h3(({
  depth
}) => ({
  color: colors.N40,
  fontSize: '0.85em',
  fontWeight: 'bold',
  marginTop: '2em',
  paddingLeft: depth ? PRIMARY_NAV_GUTTER * depth : PRIMARY_NAV_GUTTER,
  paddingRight: PRIMARY_NAV_GUTTER,
  textTransform: 'uppercase'
}));

const Wrapper = styled.nav({
  borderBottom: `1px solid ${colors.N10}`,
  color: colors.text,
  fontSize: '0.85em'
});
const SecondaryNavItem = styled(ItemElement)(({
  isSelected
}) => ({
  boxShadow: `inset 0 ${isSelected ? -2 : 0}px 0 currentColor`,
  color: isSelected ? colors.text : colors.N60,
  cursor: 'pointer',
  display: 'inline-block',
  marginRight: gridSize,
  paddingBottom: gridSize * 2,
  paddingTop: gridSize * 2,
  transition: 'box-shadow 200ms',
  ':hover': {
    color: colors.text,
    textDecoration: 'none'
  }
}));
const SecondaryNav = ({
  children
}) => /*#__PURE__*/React.createElement(Wrapper, null, children);

export { BrandItem, NavGroupIcons, NavIcon, PRIMARY_NAV_GUTTER, PrimaryNav, PrimaryNavHeading, PrimaryNavItem, PrimaryNavScrollArea, SecondaryNav, SecondaryNavItem };
