// @arch-ui/pill/dist/pill.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime/helpers/extends');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var core = require('@emotion/core');
var styled = require('@emotion/styled');
var react = require('react');
var octiconsReact = require('@primer/octicons-react');
var theme = require('@arch-ui/theme');
var common = require('@arch-ui/common');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var styled__default = /*#__PURE__*/_interopDefault(styled);

const boldBackgroundColor = {
  default: {
    default: theme.colors.N60,
    hover: theme.colors.N50,
    active: theme.colors.N70
  },
  primary: {
    default: theme.colors.primary,
    hover: theme.colors.B.L10,
    active: theme.colors.B.D10
  },
  danger: {
    default: theme.colors.danger,
    hover: theme.colors.R.L10,
    active: theme.colors.R.D10
  },
  create: {
    default: theme.colors.create,
    hover: theme.colors.G.L10,
    active: theme.colors.G.D10
  },
  warning: {
    default: theme.colors.warning,
    hover: theme.colors.Y.L10,
    active: theme.colors.Y.D10
  }
};
const boldTextColor = {
  default: 'white',
  primary: 'white',
  danger: 'white',
  create: 'white',
  warning: 'white'
};
const subtleBackgroundColor = {
  default: {
    default: theme.colors.N10,
    hover: theme.colors.N15,
    active: theme.colors.N20
  },
  primary: {
    default: theme.colors.B.L85,
    hover: theme.colors.B.L80,
    active: theme.colors.B.L75
  },
  danger: {
    default: theme.colors.R.L85,
    hover: theme.colors.R.L80,
    active: theme.colors.R.L75
  },
  create: {
    default: theme.colors.G.L85,
    hover: theme.colors.G.L80,
    active: theme.colors.G.L75
  },
  warning: {
    default: theme.colors.Y.L85,
    hover: theme.colors.Y.L80,
    active: theme.colors.Y.L75
  }
};
const subtleTextColor = {
  default: theme.colors.N50,
  primary: theme.colors.primary,
  danger: theme.colors.danger,
  create: theme.colors.create,
  warning: theme.colors.warning
};
const PillWrapper = styled__default['default'].div({
  display: 'inline-flex'
});
const PillButton = styled__default['default'].button(({
  appearance,
  variant
}) => {
  const fontSizeNumeric = 0.85;
  const fontSize = `${fontSizeNumeric}rem`;
  const borderRadius = `${fontSizeNumeric * 2}rem`;
  return _objectSpread(_objectSpread({}, common.uniformHeight), {}, {
    backgroundColor: variant === 'bold' ? boldBackgroundColor[appearance].default : subtleBackgroundColor[appearance].default,
    color: variant === 'bold' ? boldTextColor[appearance] : subtleTextColor[appearance],
    alignItems: 'center',
    border: 0,
    cursor: 'pointer',
    display: 'flex',
    fontSize: fontSize,
    justifyContent: 'center',
    maxWidth: '100%',
    minWidth: 1,
    outline: 0,
    ':hover,:focus': {
      backgroundColor: variant === 'bold' ? boldBackgroundColor[appearance].hover : subtleBackgroundColor[appearance].hover
    },
    ':active': {
      backgroundColor: variant === 'bold' ? boldBackgroundColor[appearance].active : subtleBackgroundColor[appearance].active
    },
    ':first-of-type': {
      paddingLeft: '0.9em',
      paddingRight: '0.75em',
      borderTopLeftRadius: borderRadius,
      borderBottomLeftRadius: borderRadius,
      marginRight: 1
    },
    ':last-of-type': {
      paddingLeft: '0.75em',
      paddingRight: '0.9em',
      borderTopRightRadius: borderRadius,
      borderBottomRightRadius: borderRadius,
      marginLeft: 1
    }
  });
});
const Pill = /*#__PURE__*/react.forwardRef((_ref, ref) => {
  let {
    appearance = 'default',
    children,
    onClick,
    onRemove,
    variant = 'subtle'
  } = _ref,
      props = _objectWithoutProperties(_ref, ["appearance", "children", "onClick", "onRemove", "variant"]);

  return core.jsx(PillWrapper, _extends({}, props, {
    ref: ref
  }), core.jsx(PillButton, {
    appearance: appearance,
    variant: variant,
    onClick: onClick
  }, children), onRemove ? core.jsx(PillButton, {
    appearance: appearance,
    variant: variant,
    onClick: onRemove
  }, core.jsx(octiconsReact.XIcon, {
    css: {
      height: 12
    }
  })) : null);
});

exports.Pill = Pill;
